File: backend/src/services/tts.ts
/**
 * 火山引擎 TTS (语音合成) 封装
 * - 若未配置 VOLC_TTS_APP_ID / VOLC_TTS_ACCESS_TOKEN，则返回 null
 * - 前端检测到 audioBase64 为 null 时，使用浏览器内置 SpeechSynthesis 回退
 * - 返回 base64 编码的音频数据
 */

import fetch from 'node-fetch';

export interface TTSResult {
  audioBase64: string;
  mimeType: string;
}

export interface TTSSettings {
  voiceType?: string;
  speedRatio?: number; // 0.5 - 2.0
  volumeRatio?: number; // 0.1 - 3.0
}

/** 是否配置了有效的火山引擎 TTS 凭证 */
export function isTTSConfigured(): boolean {
  const appId = process.env.VOLC_TTS_APP_ID;
  const token = process.env.VOLC_TTS_ACCESS_TOKEN;
  return !!(appId && token);
}

/**
 * 调用火山引擎 TTS。若未配置凭证或调用失败，返回 null（让前端走浏览器内置 TTS）。
 */
export async function synthesize(
  text: string,
  settings?: TTSSettings,
): Promise<TTSResult | null> {
  if (!text || !text.trim()) return null;
  if (!isTTSConfigured()) {
    console.log('[tts] 未配置火山引擎 TTS 凭证，跳过 TTS，前端将使用浏览器内置 TTS 回退。');
    return null;
  }

  const appId = process.env.VOLC_TTS_APP_ID as string;
  const accessToken = process.env.VOLC_TTS_ACCESS_TOKEN as string;
  const cluster = process.env.VOLC_TTS_CLUSTER || 'volcano_tts';
  const voiceType =
    settings?.voiceType || process.env.VOLC_TTS_VOICE_TYPE || 'BV001_streaming';

  // 截断超长文本，避免 TTS 失败
  const inputText = text.length > 1024 ? text.slice(0, 1024) : text;

  const body = {
    app: {
      appid: appId,
      token: accessToken,
      cluster,
    },
    user: {
      uid: 'ai-talking-user',
    },
    audio: {
      voice_type: voiceType,
      encoding: 'mp3',
      speed_ratio: settings?.speedRatio ?? 1.0,
      volume_ratio: settings?.volumeRatio ?? 1.0,
      pitch_ratio: 1.0,
    },
    request: {
      reqid: `req_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`,
      text: inputText,
      text_type: 'plain',
      operation: 'query',
    },
  };

  try {
    console.log(
      `[tts] 调用火山 TTS: voice=${voiceType}, textLen=${inputText.length}`,
    );
    const resp = await fetch('https://openspeech.bytedance.com/api/v1/tts', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer;${accessToken}`,
      },
      body: JSON.stringify(body),
    });

    if (!resp.ok) {
      const errText = await resp.text();
      console.warn(`[tts] 失败 ${resp.status}: ${errText.slice(0, 200)}`);
      return null;
    }

    const data = (await resp.json()) as any;
    const base64: string | undefined = data?.data;
    if (!base64) {
      console.warn('[tts] 响应中没有 data 字段');
      return null;
    }
    console.log(`[tts] 合成完成, base64 len=${base64.length}`);
    return {
      audioBase64: base64,
      mimeType: 'audio/mpeg',
    };
  } catch (err) {
    console.warn('[tts] 调用异常:', (err as Error).message);
    return null;
  }
}

File: backend-python/app/services/tts.py
import json
import time
from dataclasses import dataclass

import httpx
from loguru import logger

from app.config.settings import get_settings


@dataclass
class TTSResult:
    audio_base64: str
    mime_type: str


class TTSService:

    def __init__(self):
        self._configured: bool | None = None

    def is_configured(self) -> bool:
        if self._configured is None:
            cfg = get_settings()
            self._configured = bool(cfg.volc_tts_app_id and cfg.volc_tts_access_token)
        return self._configured

    async def synthesize(
        self,
        text: str,
        voice_type: str | None = None,
    ) -> TTSResult | None:
        if not text or not text.strip():
            return None
        if not self.is_configured():
            logger.info("[tts] 未配置 TTS 凭证，跳过。前端将使用浏览器内置 TTS。")
            return None

        cfg = get_settings()
        input_text = text[:1024]
        voice = voice_type or cfg.volc_tts_voice_type or "BV001_streaming"

        body = {
            "app": {
                "appid": cfg.volc_tts_app_id,
                "token": cfg.volc_tts_access_token,
                "cluster": cfg.volc_tts_cluster or "volcano_tts",
            },
            "user": {"uid": "ai-talking-user"},
            "audio": {
                "voice_type": voice,
                "encoding": "mp3",
                "speed_ratio": 1.0,
                "volume_ratio": 1.0,
                "pitch_ratio": 1.0,
            },
            "request": {
                "reqid": f"req_{int(time.time() * 1000)}_{int(time.time() * 1000) % 1000000}",
                "text": input_text,
                "text_type": "plain",
                "operation": "query",
            },
        }

        try:
            logger.info(f"[tts] 调用火山 TTS: voice={voice}, textLen={len(input_text)}")
            async with httpx.AsyncClient(timeout=30.0) as client:
                resp = await client.post(
                    "https://openspeech.bytedance.com/api/v1/tts",
                    headers={
                        "Content-Type": "application/json",
                        "Authorization": f"Bearer;{cfg.volc_tts_access_token}",
                    },
                    json=body,
                )
                if resp.status_code != 200:
                    logger.warning(f"[tts] 失败 {resp.status_code}: {resp.text[:200]}")
                    return None

                data = resp.json()
                base64_audio = data.get("data")
                if not base64_audio:
                    logger.warning("[tts] 响应中没有 data 字段")
                    return None

                logger.info(f"[tts] 合成完成, base64 len={len(base64_audio)}")
                return TTSResult(audio_base64=base64_audio, mime_type="audio/mpeg")

        except Exception as e:
            logger.warning(f"[tts] 调用异常: {e}")
            return None


tts_service = TTSService()
